abi_decode.go

```diff
@@ -0,0 +1,427 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+package hedera
+
+import (
+	"encoding/binary"
+	"encoding/hex"
+	"fmt"
+	"math/big"
+	"reflect"
+	"strconv"
+	"strings"
+
+	"github.com/mitchellh/mapstructure"
+)
+
+// Decode decodes the input with a given type
+func Decode(t *Type, input []byte) (interface{}, error) {
+	if len(input) == 0 {
+		return nil, fmt.Errorf("empty input")
+	}
+	val, _, err := decode(t, input)
+	return val, err
+}
+
+// DecodeStruct decodes the input with a type to a struct
+func DecodeStruct(t *Type, input []byte, out interface{}) error {
+	val, err := Decode(t, input)
+	if err != nil {
+		return err
+	}
+
+	dc := &mapstructure.DecoderConfig{
+		Result:           out,
+		WeaklyTypedInput: true,
+		TagName:          "abi",
+	}
+	ms, err := mapstructure.NewDecoder(dc)
+	if err != nil {
+		return err
+	}
+	if err = ms.Decode(val); err != nil {
+		return err
+	}
+	return nil
+}
+
+func decode(t *Type, input []byte) (interface{}, []byte, error) {
+	var data []byte
+	var length int
+	var err error
+
+	// safe check, input should be at least 32 bytes
+	if len(input) < 32 {
+		return nil, nil, fmt.Errorf("incorrect length")
+	}
+
+	if t.isVariableInput() {
+		length, err = readLength(input)
+		if err != nil {
+			return nil, nil, err
+		}
+	} else {
+		data = input[:32]
+	}
+
+	switch t.kind {
+	case KindTuple:
+		return decodeTuple(t, input)
+
+	case KindSlice:
+		return decodeArraySlice(t, input[32:], length)
+
+	case KindArray:
+		return decodeArraySlice(t, input, t.size)
+	}
+
+	var val interface{}
+	switch t.kind {
+	case KindBool:
+		val, err = decodeBool(data)
+
+	case KindInt, KindUInt:
+		val = readInteger(t, data)
+
+	case KindString:
+		val = string(input[32 : 32+length])
+
+	case KindBytes:
+		val = input[32 : 32+length]
+
+	case KindAddress:
+		val, err = readAddr(data)
+
+	case KindFixedBytes:
+		val, err = readFixedBytes(t, data)
+
+	case KindFunction:
+		val, err = readFunctionType(t, data)
+
+	default:
+		return nil, nil, fmt.Errorf("decoding not available for type '%s'", t.kind)
+	}
+
+	return val, input[32:], err
+}
+
+var (
+	maxUint256 = big.NewInt(0).Add(
+		big.NewInt(0).Exp(big.NewInt(2), big.NewInt(256), nil),
+		big.NewInt(-1))
+	maxInt256 = big.NewInt(0).Add(
+		big.NewInt(0).Exp(big.NewInt(2), big.NewInt(255), nil),
+		big.NewInt(-1))
+)
+
+// Address is an Ethereum address
+type Address [20]byte
+
+func min(i, j int) int {
+	if i < j {
+		return i
+	}
+	return j
+}
+
+// BytesToAddress converts bytes to an address object
+func BytesToAddress(b []byte) Address {
+	var a Address
+
+	size := len(b)
+	min := min(size, 20)
+
+	copy(a[20-min:], b[len(b)-min:])
+	return a
+}
+
+// Address implements the ethgo.Key interface Address method.
+func (a Address) Address() Address {
+	return a
+}
+
+// Sign implements the ethgo.Key interface Sign method.
+func (a Address) Sign(hash []byte) ([]byte, error) {
```

Comment:
If `an address cannot sign messages` why have this function?

Reply:
removed.